### DISCUSSION TITLE
Not able to reset password through PKCE flow in NextJS

### DISCUSSION THREAD

**Discussion by kartikguptaofficial** [supabase/supabase#28655] [Questions] [answered]
I am using NextJS (App Router) as the backend and react native as the frontend, I am trying to create authentication APIs in the NextJS which will be integrated to the react native mobile app.
The APIs are for : signin, signup, forgot password and reset password.
When I implemented these four APIs in the nodejs (express) it worked fine and also the reset-password flow worked perfectly which used the `exhangeCodeForSession` method which returns the access token and refresh token, and after receiving these tokens I can set the session and `updateUser` password.
But, when I tried the same flow in nextjs, the forgot password API which uses the `resetPasswordForEmail` method sends the reset password email, but when the callback route is called with the verification code, the `exhangeCodeForSession` returns an error i.e. `invalid request: both auth code and code verifier should be non-empty`, I am not getting a way to resolve this, as I don't know much about the working of this flow and how this code verifier works and how to get this.

Please help me in finding the solution for this, or how I can achieve this flow in nextjs.. as in nodejs it is working (also please guide me how it is different from nextjs).

I have implemented the supabase client by looking into this document:
https://supabase.com/docs/guides/auth/server-side/nextjs

**Comment by kartikguptaofficial** [comment, marked as answer]
I got the solution for this, we have to use token_hash template for reset-password email and then make use of `verifyOtp` method to get the session on the callback URL, then we can just update the user using `updateUser` method.

Refer to this document:
https://supabase.com/docs/guides/auth/passwords?queryGroups=language&language=js&queryGroups=flow&flow=pkce&queryGroups=framework&framework=nextjs#resetting-a-password

**Reply by KaanCL** [reply]
You are a legend, thank you so much. I've been struggling with this since the morning. 🙌

**Reply by satyamneo00** [reply]
I have been searching for a solution to this since morning. Thanks bro
